### DISCUSSION TITLE
Cec Module Parameters

### DISCUSSION THREAD

**Discussion by NWethe** [pvlib/pvlib-python#1560] [Q&A]
Hey all,

I would like to know where the values for the modules in the Cec Database are derived from. I'm interested in adding modules into my code myself, which I would like to use to calculate the AC power in the end. A problem I am running into: 

```
"""PV module data from a typical datasheet (e.g. Kyocera Solar KD225GX LPB)."""
module_data = {'celltype': 'multiSi', # technology
               'v_mp': 29.8, # Maximum power voltage
               'i_mp': 7.55, # Maximum power current
               'v_oc': 36.9, # Open-circuit voltage
               'i_sc': 8.18, # Short-circuit current
               'alpha_sc': 0.001636, # Temperature Coeff. Short Circuit Current [A/C]
               'beta_voc': -0.12177, # Temperature Coeff. Open Circuit Voltage [V/C]
               'gamma_pmp': -0.43, # Temperature coefficient of power at maximum point [%/C]
               'cells_in_series': 96, # Number of cells in series
               'temp_ref': 25}  # Reference temperature conditions
```

these parameters were given for the module Kyocera Solar KD225GX LPB in the Cec Database. My cods runs smoothly with these given parameters. Though now I noticed that, if I google datasheets for Kyocera Solar KD225GX LPB the parameters are the same, besides Alpha_sc, Beta_voc and Gamma_pmp. In online datasheets they are given as:

Alpha_sc : 4.91*10^(-3) [A/C]
Beta_voc : -0,133 [V/C]
Gamma_pmp : -0.46 [%/C]

If I insert these values instead, I obviously get an error message when I want to use the Cec Model for the single diode equation. Does anyone of you know how the values for the temperature coefficients in the Cec Database are calculated - and if it's possible to add modules on your own, just by looking up their datasheets?

I hope my question is understandable and thank you very much in advance!

**Comment by cwhanse** [comment]
The CEC model parameters are computed by the SAM team as part of their process for new SAM versions. We (pvlib and SAM) are in agreement that the code and procedures should become open, but it's one of those things that neither side has been able to make time for.

For a single module, the primary tool is available through a pvlib function: 

https://github.com/pvlib/pvlib-python/blob/bdbaf4c238ce2b373d04af48193b14659662cef1/pvlib/ivtools/sdm.py#L23

You can try this function and see if you get acceptable parameters. The optimization routine is in the SAM code base and it's not guaranteed to converge. If that happens, the SAM team has a procedure for adjusting initial guesses, etc., which isn't captured in this pvlib function.

As far as the error message, can you share it? At first glance I don't see why substituting those temperature coefficient values would cause a problem. If instead, you are using that `module_data` dict as the value of the pvlib parameter `module_parameters`, the dict doesn't contain what's expected: it must have the diode model coefficients `I_L_ref` etc. that are returned by `pvlib.ivtools.fit_cec_sam`.

**Reply by kandersolar** [reply]
One thing to note is the number of cells -- 96 in @NWethe's dictionary but only 60 in SAM's CEC table (`Kyocera_Solar_KD225GX_LPB`).

**Comment by adriesse** [comment]
The SAM procedure is now available here:

https://github.com/NREL/SAM/tree/develop/samples/CEC%20Module%20and%20Inverter%20Libraries
